README.md
# Cloud Haoma

The project was done in a team effort. Team Members:-
1. Supriya Aggarwal 2014A7PS0013P
2. Vidhi Jain 2014A7TS0113P
3. Abhishek Gupta 2014A7PS0026P

Run the following commands on all the nodes of the cluster

Python version 2.7

1. sudo apt-get install nodejs nodejs-legacy npm libzmq-dev git
2. pip install pyzmq-static
2. Open Terminal in the base directory of the project
3. npm install
4. pip install zmq


Network topology is star.

Run the following on the designated router.

1. python2.7 router.py 12345 [[1,2,3],[4,5]] '[6]' false


Run the following on the cluster nodes (you may include router also (Let IPADDRESS of router node for this be 127.0.0.1):-

Group I 
1. python2.7 server.py 1 tcp://127.0.0.1 12345 ["1","2","3"] on node1
2. python2.7 server.py 2 tcp://127.0.0.1 12345 ["1","2","3"] on node2
3. python2.7 server.py 3 tcp://127.0.0.1 12345 ["1","2","3"] on node3

Group II
1. python2.7 server.py 4 tcp://127.0.0.1 12345 ["4","5"] on node4
2. python2.7 server.py 5 tcp://127.0.0.1 12345 ["4","5"] on node5



Run the following code on the node which you want to run your web server
(Currently webserver tries to insert hardcoded values in the db on receiving
 the API call instead of using request Data)

1.node clientServer.js 6 tcp://127.0.0.1 12345 true 3000



API CALL:- 


//For writing request
//Requests Information:-
//POST REQUEST - multipart/formdata
//Keys - fileData, clientId, sourceId

IPADDRESSOFWEBSERVER:3000/pushData


//For reading request
//Request Infromation:
//Post Request - multipart/formdata
//Keys - clientId, sourceId

IPADDRESSOFWEBSERVER:3000/readData




# Strategy
Using RAFT for log replication. 
Each log entry contains the write request by a client. Request contains sufficient 
information to identify the file. State machine operation is to append the request 
data to the identified file.
Reads can be handled by the leader
Data is sharded across the groups (here gropu I and II). Raft replicates the logs on all the nodes of the cluster => So if we run a single Raft instance for our cluster, we can't achieve a replication factor less than that of cluster size => Cluster should be divided into groups(shards) and raft instance should run for that group. => A data item will be present in one group only and replication is within the nodes of that group

Some consensus algorithms support partial replication in which we can control which data will replicated and its replication factor. Raft doesn't support it so we should shard our data to get decrease replication factor. Also, sharding improves writes throughput.

# Status
Code for supporting the write and read operation is working with the API Calls.

# Reference

This project is based on the raft implementation present at https://github.com/albertdb/Raft.

The entire code was rewritten into python (Original code was written in nodejs which
we suspect to be better choice than python because of its asynchronous nature)

ZeroMQ (http://zeromq.org/) is used for providing communication



changes.md
1. Small fixes
2. Make commitEntries a periodic function

clientServer.js
//node clientServer.js 5 tcp://127.0.0.1 12345 true


var app = require('express')();
var bodyParser = require('body-parser');
var formidable = require('express-formidable');
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended:true}));
app.use(formidable());
var http = require('http').Server(app);
var zmq = require('zmq');
require('events').EventEmitter.defaultMaxListeners = 100000000;

serverID = process.argv[2]
router_address = process.argv[3]
send_port_no = process.argv[4]
debug = (process.argv[5] == "true")

listen_socket = zmq.socket('pull');
listen_socket.connect(router_address+":5002")

sender_socket = zmq.socket('push')
sender_socket.connect(router_address+":"+send_port_no)

var writePending_dict = [],
readPending_dict = [];


listen_socket.on('message', function(){

  // Parsing Message

  var args = Array.apply(null, arguments);
  console.log(args[0].toString('utf8'));
  var message = JSON.parse(args[0].toString('utf8'))
  console.log(message);


  if (message['rpc'] == 'addEntryReply')
  	writePending_dict[message['requestId']] = message['status'];
  else if (message['rpc'] == 'readEntryReply'){
  	readPending_dict[message['requestId']] = message['data'];
  }

});


//For reading request
//Request Infromation:
//Post Request - multipart/formdata
//Keys - clientId, sourceId
app.post('/readData', function(req, res){
  var requestId = Math.random().toString(36).substring(7); //Assigning a randomID
  if (debug) console.log("Read RequestID assigned : " + requestId)
  readPending_dict[requestId] = 'failure';
  var data = req.fields;
  if (debug) console.log(data);
  var fileName = data['clientId']+"/"+data['sourceId'];
  msg = {'clientId' : serverID
  , 'dest' : 1
  , 'requestId': requestId
  , 'rpc' : 'readEntry'
  , 'fileName' : fileName
  , 'customerId' : data['clientId']
  , 'sourceId' : data['sourceId']
  };
  sender_socket.send(JSON.stringify(msg));
  if (debug) console.log(msg);
  var count = 20;
  var my_int = setInterval(function(){
    if (count == 0){
      res.send("Server Timeout!");
      if (debug) "Request id " + requestId + "didn't got serve\n"
      clearInterval(my_int);
      return;
    }
    count -= 1;
    if (readPending_dict[requestId] != "failure"){
      if (debug) console.log(readPending_dict[requestId]);
      res.send(readPending_dict[requestId]);
      clearInterval(my_int);
    }
  }, 100);
  return;
});


//For writing request
//Requests Information:-
//POST REQUEST - multipart/formdata
//Keys - fileData, clientId, sourceId
app.post('/pushData', function(req, res){
	var requestId = Math.random().toString(36).substring(7); //Assigning a randomID
	if (debug) console.log("Write RequestID assigned : " + requestId)
	writePending_dict[requestId] = 'failure';
  data = req.fields;
  if (debug) console.log(data)
  msg = {'clientId':serverID
  , 'requestId' : requestId
  , 'requestData' : {'fileData': new Date().getTime() + ":" + data['fileData'] + "\n", 'clientId': data['clientId'], 'sourceId':data['sourceId'] }
  , 'rpc':'addEntry'
  , 'customerId' : data['clientId']
  , 'sourceId' : data['sourceId']
  };
  sender_socket.send(JSON.stringify(msg));
  if (debug) console.log(msg)
  var count = 20;
  var my_int =   setInterval(function(){
    if (count == 0) {
    	res.send("Server Timeout!");
    	clearInterval(my_int);
  	  return;
    }
    count -= 1;
    if (writePending_dict[requestId] == "Success"){
  	  if (debug) console.log("Log accepted");
  	  res.send("Log added!");
  	  clearInterval(my_int);
    }
  },100);
  return;
});


//Start http server
http.listen(3000, function(){
  console.log('listening on *:3000');
});


